gallery/gallery/lib/studies/shrine/supplemental/balanced_layout.dart

```diff
@@ -0,0 +1,305 @@
+// Copyright 2019 The Flutter team. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'package:gallery/studies/shrine/model/product.dart';
+import 'package:gallery/studies/shrine/supplemental/desktop_product_columns.dart';
+import 'package:gallery/studies/shrine/supplemental/layout_cache.dart';
+
+/// A placeholder id for an empty element. See [_iterateUntilBalanced]
+/// for more information.
+const _emptyElement = -1;
+
+/// To avoid infinite loops, improvements to the layout are only performed
+/// when a column's height changes by more than
+/// [_deviationImprovementThreshold] pixels.
+const _deviationImprovementThreshold = 10;
+
+/// Height of a product image, paired with the product's id.
+class _TaggedHeightData {
+  const _TaggedHeightData({
+    @required this.index,
+    @required this.height,
+  });
+
+  /// The id of the corresponding product.
+  final int index;
+
+  /// The height of the product image.
+  final double height;
+}
+
+/// Converts a set of [_TaggedHeightData] elements to a list,
+/// and add an empty element.
+/// Used for iteration.
+List<_TaggedHeightData> toListAndAddEmpty(Set<_TaggedHeightData> set) {
+  List<_TaggedHeightData> result = List<_TaggedHeightData>.from(set);
+  result.add(_TaggedHeightData(index: _emptyElement, height: 0));
+  return result;
+}
+
+/// Sort a list of integers, then return the sorted list.
+/// Used in list comprehensions using "for".
+List<int> _sorted(List<int> list) {
```

Comment:
Now that extension methods are part of flutter stable, this could probably be done as an extension on List instead

Reply:
@rami-a Thank you! `flutter analyze` requires me to update SDK to 2.6.0 or later. Shall I?

```
   info • Extension methods weren't supported until version 2.6.0, but this code is required to be able to run on earlier versions •
          lib/studies/shrine/supplemental/balanced_layout.dart:45:1 • sdk_version_extension_methods
```

This requires the following version in `pubspec.yaml`:

```
environment:
  sdk: ">=2.6.0 <3.0.0"
```

Reply:
Is this method necessary? It looks like it's just wrapping `sort`.

Reply:
I think this should be fine to do since flutter stable is now on dart 2.6. As long as the analyzer on this repo is happy then go ahead and do it!

Reply:
@rami-a @RedBrogdon 

I just realized I can use `list..sort()` to sort and return the list. Thank you for the suggestion!